File: Excel/623 Challenge.py
import pandas as pd
import numpy as np

path = "623 Sum of Numbers Across Diagonals of 3 Columns.xlsx"
input_data = pd.read_excel(path, usecols="A:C", nrows=19).values
test = pd.read_excel(path, usecols="E", nrows=1).squeeze()

def construct_zigzag(n, width=3):
    return np.tile(np.concatenate([np.arange(1, width + 1), np.arange(width - 1, 1, -1)]), n)[:n]

zigzag = construct_zigzag(len(input_data))
result = np.sum(input_data[np.arange(len(input_data)), zigzag - 1])

print(result == test) # True

File: Excel/600-699/623/623 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/623 Sum of Numbers Across Diagonals of 3 Columns.xlsx"
input = read_excel(path, range = "A1:C19") %>% as.matrix()
test  = read_excel(path, range = "E1:E2") %>% pull()


construct_zigzag <- function(n, width = 3) {
  repeating_block <- c(1:width, (width - 1):2)
  rep_len(repeating_block, n)
}

zigzag = construct_zigzag(nrow(input))
result = sum(input[cbind(1:nrow(input), zigzag)])

test == result
#> [1] TRUE
